File: document-workspace-backend/src/main/java/com/pablodev/documentworkspace/model/Folder.java
package com.pablodev.documentworkspace.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.SuperBuilder;

import java.util.List;

@Entity
@Table(name = "folders",
        uniqueConstraints = @UniqueConstraint(
                columnNames = {"name", "parent_folder_id"}
        )
)
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@SuperBuilder
public class Folder extends AbstractEntity  {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String name;

    @ManyToOne
    private Folder parentFolder;

    @OneToMany(mappedBy = "folder", cascade = CascadeType.ALL)
    private List<Document> documents;

    @OneToMany(mappedBy = "parentFolder", cascade = CascadeType.ALL)
    private List<Folder> subFolders;

}

File: document-workspace-frontend/src/app/core/model/folder/folder.ts
import {DocumentResponse} from "../document/document-response";
import {FolderInfoResponse} from "./folder-info-response";

export interface Folder {
    id: number;
    name: string;
    createdDate: string;
    lastModifiedDate: string;
    documents: DocumentResponse[];
    folders: FolderInfoResponse[];
    parentFolder: FolderInfoResponse;
}
